File: day6/puzzle1.lua
local function main()
    local f = io.open("puzzle_input.txt", "r")
    if not f then
        os.exit(1)
    end
    local times = {}
    local distances = {}
    for line in f:lines() do
        if line:sub(1, 5) == "Time:" then
            for time in line:sub(6, -1):gmatch("%d+") do
                table.insert(times, tonumber(time))
            end
        elseif line:sub(1, 9) == "Distance:" then
            for distance in line:sub(10, -1):gmatch("%d+") do
                table.insert(distances, tonumber(distance))
            end
        end
    end
    f:close()
    local travels = {}
    for i = 1, #times do
        local options = 0
        for hold_time = 0, times[i] - 1 do
            local speed = hold_time
            local distance = (times[i] - hold_time) * speed
            if distance > distances[i] then
                options = options + 1
            end
        end
        table.insert(travels, options)
    end
    local margin_of_error = 1
    for _, travel in ipairs(travels) do
        margin_of_error = margin_of_error * travel
    end
    print(string.format("margin_of_error = %d", margin_of_error))
end

main()

File: day6/puzzle1.py
from re import compile


def main():
    pattern = compile(r"\d+")
    with open("puzzle_input.txt", "rt") as f:
        for line in f:
            if line.startswith("Time:"):
                times = [int(t) for t in pattern.findall(line)]
            elif line.startswith("Distance:"):
                distances = [int(t) for t in pattern.findall(line)]
            else:
                continue
    travels: list[int] = []
    for i in range(len(times)):
        options = 0
        for hold_time in range(times[i]):
            speed = hold_time
            distance = (times[i] - hold_time) * speed
            if distance > distances[i]:
                options += 1
        travels.append(options)
    margin_of_error = 1
    for travel in travels:
        margin_of_error *= travel
    print(f"margin of error = {margin_of_error}")


if __name__ == "__main__":
    main()
